### ISSUE TITLE
Bug with KeyFrame constructors

### ISSUE THREAD

**Issue by geokon-gh** [halgari/fn-fx#34]
When inspecting the constructors of KeyFrame, it returns a default/empty constructor:
ie:
`(fn-fx.render-core/get-value-ctors javafx.animation.KeyFrame);javafx.scene.image.Image)
`
returns 
`{[] #function[fn-fx.render-core/get-value-ctors/iter--812--816/fn--817/fn--826]}`

While it should have multiple constructor values (:time and :on-finished) https://docs.oracle.com/javase/8/javafx/api/javafx/animation/KeyFrame.html

**Comment by geokon-gh** [comment]
Okay, I figured out where the issue is. In `fn-fx.util.reflect-utils/get-arg-names` the part that gets the `params` is having issues

0: 
``` clojure
(.getType (nth  (.getParameters (:method (first (fn-fx.util.reflect-utils/get-ctors javafx.animation.KeyFrame)))) 0 ))
```
will return
``` clojure
javafx.util.Duration
```


1: 
``` clojure
(.getType (nth  (.getParameters (:method (first (fn-fx.util.reflect-utils/get-ctors javafx.animation.KeyFrame)))) 0 ))
```
will return
``` clojure
javafx.event.EventHandler
```



2: 
``` clojure
(.getType (nth  (.getParameters (:method (first (fn-fx.util.reflect-utils/get-ctors javafx.animation.KeyFrame)))) 0 ))
```
will return
``` clojure
[Ljavafx.animation.KeyValue;
```
and notice how this has a [L in the front and a trailing semi-colon!
According to [here](https://stackoverflow.com/questions/29587500/l-prefix-in-java-reflection) this is an encoding to say that It's an array of KeyValues

The next part I don't really get.. the file `method-arg-info.edn` has the mappings from JavaFX classes to signatures. If you look through this monster file you can fine a part that says

``` clojure
["KeyFrame" "KeyFrame" ["Duration" "EventHandler<ActionEvent>" "KeyValue"]] ["time" "onFinished" "values"]`
```
I have no idea where this files comes from, but this gives no indication that KeyValue is an array..  My guess is that there is no support for array arguments? Maybe someone can confirm?

At this point the easiest patch would be to have a check that if the param is an array, it will get its type and use that instead.

**Comment by geokon-gh** [comment]
Alright, upon further study..  in  `method-arg-info.edn` we are reading in a huge map and the `["KeyFrame" "KeyFrame" ["Duration" "EventHandler<ActionEvent>" "KeyValue"]]` part is the key and the ` ["time" "onFinished" "values"] ` is the value

The code get the function signature and tries to use it as a key to get the values (which are the parameter names.. I guess they need to be preprocesses and shoved into this extra file b/c they can't be determined at runtime or something)

The line 

``` clojure 
(map #(-> % .getType .getName fn-fx.util.reflect-utils/unqualified-name) (.getParameters (:method (first (fn-fx.util.reflect-utils/get-ctors javafx.animation.KeyFrame)))))
```
returns 
```
("Duration" "EventHandler" "KeyValue;")
```

Which I guess has two problems. The `KeyValue;` with the comma.. and the `EvenHandler` without the `<ActionEvent>` extra type info. The comma seems to disappear after it's all put together into a map in `[class-name method-name params]` 
and we get `["KeyFrame" "KeyFrame" ("Duration" "EventHandler" "KeyValue")]` which doesn't work as a key!

But `["KeyFrame" "KeyFrame" ["Duration" "EventHandler<ActionEvent>" "KeyValue"]])` does..

Backpedaling a bit, we need to look at what we get when we actually ask the system for the signature.. 
Running for instance `(first (.getConstructors javafx.animation.KeyFrame))` we get `#object[java.lang.reflect.Constructor 0x12c1dc7c "public javafx.animation.KeyFrame(javafx.util.Duration,javafx.event.EventHandler,javafx.animation.KeyValue[])"]` Which has nothing about the `<ActionEvent>`. So my guess is that the real issue is with the `method-arg-info.edn` file and it needs to be fixed there!

**Comment by geokon-gh** [comment]
I see this project has come back from the dead. I'm available to help debug this..

**Comment by pmonks** [comment]
@geokon-gh fantastic!  Would you mind checking out the latest source, and reproducing your issue and seeing if it still occurs, and if so whether the symptoms are the same?

We hope to have automatic SNAPSHOT builds deployed to Clojars soon (see issue #43 and PR that addresses it), but that fix is still pending as of right now.

**Comment by pmonks** [comment]
Yeah sorry for the poor developer experience.  Once we get a 0.5.0-SNAPSHOT out (i.e. fix issue #43) it should hopefully improve!  For now the way I would suggest testing is to clone the repo, install the JAR locally, then start a REPL.

To clone & install the JAR locally:

```shell
$ cd <my local development directory>
$ git clone https://github.com/fn-fx/fn-fx.git
$ cd fn-fx
$ lein install
$ cd ..
```

To start a [`lein-try`](https://github.com/avescodes/lein-try) REPL:

```shell
$ lein try fn-fx/fn-fx "0.5.0-SNAPSHOT-openjfx11   # If on JDK 11
$ lein try fn-fx/fn-fx "0.5.0-SNAPSHOT-javafx         # If on JDK 7-10
```

Or, if you'd rather not use `lein-try`, you could create a new folder and put a `project.clj` file in it, something like this:

```clojure
(defproject geokon-gh/my-fn-fx-project "0.1.0-SNAPSHOT"
  :description      "My fn-fx project"
  :min-lein-version "2.8.1"
  :dependencies     [[org.clojure/clojure         "1.9.0"]
                     [fn-fx/fn-fx "0.5.0-SNAPSHOT"]])
```

Then start a vanilla REPL:

```shell
$ lein repl
nREPL server started on port 55554 on host 127.0.0.1 - nrepl://127.0.0.1:55554
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.9.0
OpenJDK 64-Bit Server VM 11.0.1+13
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

user=> 
```

Either way, once you're in a REPL you should be able to get started on reproducing your issue via standard `fn-fx` API calls:

```clojure
user=> (require '[fn-fx.fx-dom :as dom])
nil
user=> (require '[fn-fx.controls :as ui])
nil
user=> ; Start reproing here
```

I wish I could help you more with the `fn-fx` specific parts here, but I'm a complete n00b with the library and JavaFX myself.  In fact my selfish motivation in helping to resurrect the project is to learn it!  😉

**Comment by geokon-gh** [comment]
thanks for those tips @pmonks! Always a bit hard to find the right workflow for things like this.

My first 2 posts kinda illustrate as far as I can understand the picture - but I hit a bit of wall after that. Hopefully you can more sense of things

**Comment by geokon-gh** [comment]
Okay, now that I have things working again, maybe I can take a stab at this again :)

Have you been able to regenerate the `method-arg-info.edn` file @pmonks ? And.. have there not been any changes to the JavaFX API between Java 8 and 11? B/c I'm seeing just one `method-arg-info.edn`
